fcrepo4/fcrepo4
# Remove ocfl id mappings on purge

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3357

Original PR: https://github.com/fcrepo4/fcrepo4/pull/1710

# What does this Pull Request do?

1. OCFL ID mapping changes are accumulated in a transaction table and only applied when a transaction is committed
1. When resources are purged their associated mappings are removed from the OCFL ID map table

# How should this be tested?

1. Start Fedora against a DB like Postgres.

2. Fire up the DB's CLI and execute: `select * from ocfl_id_map;`

3. Create resources:

```
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/ag-02 -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -H'Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG Parent 2'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/ag-02/child -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG 2 Child'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/ag-02/child/grand -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG 2 grandchild'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/ag-02/child/grand/great -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG 2 great grandchild'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/ag-02/child/grand/great/foo -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "bar"
```

4. Verify the mappings in the DB

5. Purge a resource:

```
curl -u fedoraAdmin:fedoraAdmin -v -XDELETE http://localhost:8080/rest/ag-02/child
curl -u fedoraAdmin:fedoraAdmin -v -XDELETE http://localhost:8080/rest/ag-02/child/fcr:tombstone
```

6. Verify the mapping, and all child mappings, were removed from the DB

7. Create a new resource:

```
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/foo -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "bar"
```

8. Purge the resource within a transaction (DO NOT COMMIT):

```
curl -u fedoraAdmin:fedoraAdmin -v -XPOST http://localhost:8080/rest/fcr:tx
txid=ebeb046a-64e0-4e46-9ed6-38b3cd1e20a2
curl -u fedoraAdmin:fedoraAdmin -v -H"Atomic-ID: ${txid}" -XDELETE http://localhost:8080/rest/foo
curl -u fedoraAdmin:fedoraAdmin -v -H"Atomic-ID: ${txid}" -XDELETE http://localhost:8080/rest/foo/fcr:tombstone
```

9. Verify the resource still exists in the DB and is access outside the transaction, but not within:

```
curl -u fedoraAdmin:fedoraAdmin -H"Atomic-ID: ${txid}" -v http://localhost:8080/rest/foo

curl -u fedoraAdmin:fedoraAdmin -v http://localhost:8080/rest/foo
```

10. Commit the transaction:

```
curl -u fedoraAdmin:fedoraAdmin -X PUT -v http://localhost:8080/rest/fcr:tx/${txid}
```

11. Verify the resource no longer exists in the DB and can no longer be accessed

# Interested parties
@fcrepo4/committers
